pkg/tbtc/signing.go

```diff
@@ -2,378 +2,372 @@ package tbtc
 
 import (
 	"context"
-	"crypto/sha256"
-	"encoding/binary"
 	"fmt"
-	"math/big"
-	"math/rand"
-	"sort"
-
-	"github.com/ipfs/go-log/v2"
-	"github.com/keep-network/keep-core/pkg/chain"
+	"github.com/keep-network/keep-core/pkg/generator"
+	"github.com/keep-network/keep-core/pkg/net"
+	"github.com/keep-network/keep-core/pkg/protocol/announcer"
 	"github.com/keep-network/keep-core/pkg/protocol/group"
-	"github.com/keep-network/keep-core/pkg/tecdsa/retry"
+	"github.com/keep-network/keep-core/pkg/tecdsa"
 	"github.com/keep-network/keep-core/pkg/tecdsa/signing"
+	"go.uber.org/zap"
 	"golang.org/x/exp/slices"
+	"golang.org/x/sync/semaphore"
+	"math/big"
+	"reflect"
+	"strings"
 )
 
-// signingAttemptMaxBlockDuration determines the maximum block duration of a
-// single signing attempt.
-const signingAttemptMaxBlockDuration = 100
-
-// signingAnnouncer represents a component responsible for exchanging readiness
-// announcements for the given signing attempt of the given message.
-type signingAnnouncer interface {
-	Announce(
-		ctx context.Context,
-		memberIndex group.MemberIndex,
-		sessionID string,
-	) ([]group.MemberIndex, error)
-}
-
-// signingRetryLoop is a struct that encapsulates the signing retry logic.
-type signingRetryLoop struct {
-	logger log.StandardLogger
-
-	message *big.Int
-
-	signingGroupMemberIndex group.MemberIndex
-	signingGroupOperators   chain.Addresses
-
-	chainConfig *ChainConfig
-
-	announcer                signingAnnouncer
-	announcementDelayBlocks  uint64
-	announcementActiveBlocks uint64
+const (
+	// signingBatchInterludeBlocks determines the block duration of the
+	// interlude preserved between subsequent signings in a signing batch.
+	signingBatchInterludeBlocks = 2
```

Comment:
Suppose we have two messages to sign, A and B, and 5 signers. Let's say the signature of A is computed in the following blocks:
- Member 1 computes it at block 100
- Member 2 computes it at block 101
- Member 3 computes it at block 102
- Member 4 computes it at block 103
- Member 5 computes it at block 104

Each signer starts the done check communication as soon as it has the signature, so member 1 at block 100, member 2 at block 101, and so on. The slowest signer is member 5 so their end block (104) becomes the common end block used by the entire group. As soon as members receive the expected done checks, their return (but still broadcast their done checks until the attempt timeout) and move to the next message (B) in the batch. If we don't use the interlude, the signing of message B is supposed to start at block 104 when none of the signers are actually ready to begin it yet, as they are still executing the previous signature for message A. Nothing bad happens as they will catch up eventually but we lose some time intended for the first signing attempt of B. If we add the interlude, we basically give some time to complete the signature of A before starting B and the attempt's time loss is smaller. 

Note that we can't determine a strict interlude duration that will match all cases well but we actually don't have to. Two blocks (~25 seconds) seem to be a good approximation based on the test results to work in most cases.

Reply:
I need to look at the code one more time as there is something I do not understand here. I would appreciate explaining it in English one more time so that I can confront the expectations.

> As soon as members receive the expected done checks, their return (but still broadcast their done checks until the attempt timeout) and move to the next message (B) in the batch.

Don't they stop broadcasting the done signal when they receive the done signal from the rest of the group? This feels logic. If everyone agreed the work is done, there is no one left behind, and retransmitting a done signal or any other message from the given signing session is no longer needed.

>  If we don't use the interlude, the signing of message B is supposed to start at block 104 when none of the signers are actually ready to begin it yet, as they are still executing the previous signature for message A. 

Why are they not ready? They received done signal from everyone else, they should stop the signing and move to the next message. Sure, they need some time to do that but we execute the announcement phase for each signing separately and there is a one-block delay at the beginning.

Reply:
> Don't they stop broadcasting the done signal when they receive the done signal from the rest of the group? This feels logic. If everyone agreed the work is done, there is no one left behind, and retransmitting a done signal or any other message from the given signing session is no longer needed.

If member 1 received all expected done signals, that only means other peers that took part in signing sent their done checks and they reached member 1. This says nothing about whether the other peers received all the required checks too. This is why member 1 must keep retransmitting its done check until the attempt timeout (this is a very similar case than described in https://github.com/keep-network/keep-core/pull/3392#discussion_r10142110310)

> Why are they not ready? They received done signal from everyone else, they should stop the signing and move to the next message. Sure, they need some time to do that but we execute the announcement phase for each signing separately and there is a one-block delay at the beginning.

The end block does not say anything about the done check phase. It says "signature was computed at block X", point.

Reply:
That all makes perfect sense. I suggest improving the documentation of `signingBatchInterludeBlocks` and `attemptCtx` (in a separate comment):
```
	// signingBatchInterludeBlocks determines the block duration of the
	// interlude preserved between subsequent signings in a signing batch.
	// If the signing of the previous message completed at block X, the signing 
	// of the next message starts at `X + signingBatchInterludeBlocks`.
	// This is the additional time signers have to realize that the signing is
	// done by receiving the signingDoneMessage. Note that the end block of the
	// previous signing used to establish the start block of the next signing
	// comes from signingDoneMessage received and there is no guarantee all
	// signing group members received signingDoneMessage before the highest
	// endBlock is reached on the chain. The interlude is an additional time for 
	// the broadcast channel to spread information about signing successfully
	// completed by the slowest signing group member (the one who sends the
	// signingDoneMessage as the last one).
	signingBatchInterludeBlocks = 2
```

Reply:
See 429d7cde and 81f1c4ff

Reply:
See 429d7cde and 81f1c4ff